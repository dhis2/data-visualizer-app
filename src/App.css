.visualization-type-selector {
    grid-area: charttype;
}
.menu-bar {
    grid-area: menubar;
}
.dimensions {
    grid-area: dimensions;
}
.canvas {
    grid-area: canvas;
}

.app {
    display: grid;
    grid-template-areas:
        'charttype menubar'
        'dimensions canvas';
    grid-template-columns: 250px auto;
    grid-template-rows: 40px auto;
    grid-gap: 5px;
    padding: 5px;
    font-family: 'roboto';
}
