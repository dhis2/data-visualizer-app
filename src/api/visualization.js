import { getFieldsStringByType } from '../modules/fields/index.js'

const visualizationQuery = {
    visualization: {
        resource: 'visualizations',
        id: ({ id }) => id,
        params: ({ withSubscribers }) => ({
            fields: getFieldsStringByType('visualization', { withSubscribers }),
        }),
    },
}

const visualizationNameDescQuery = {
    visualization: {
        resource: 'visualizations',
        id: ({ id }) => id,
        params: {
            fields: 'name,displayName,description,displayDescription',
        },
    },
}

const visualizationSubscribersQuery = {
    visualizationSubscribers: {
        resource: 'visualizations',
        id: ({ id }) => id,
        params: {
            fields: ['subscribers'],
        },
    },
}

const visualizationsQuery = {
    visualization: {
        resource: 'visualizations',
        params: ({ visualizationIds }) => ({
            filter: `id:in:[${visualizationIds}]`,
            fields: ['id', 'type'],
        }),
    },
}

export const apiFetchVisualization = ({
    engine,
    id,
    withSubscribers = false,
}) => {
    return engine.query(visualizationQuery, {
        variables: { id, withSubscribers },
    })
}

export const apiFetchVisualizationNameDesc = (engine, id) => {
    return engine.query(visualizationNameDescQuery, {
        variables: { id },
    })
}

export const apiFetchVisualizationSubscribers = (engine, id) => {
    return engine.query(visualizationSubscribersQuery, {
        variables: { id },
    })
}

export const apiFetchVisualizations = (engine, visualizationIds) => {
    return engine.query(visualizationsQuery, {
        variables: { visualizationIds },
    })
}

export const apiSaveVisualization = (engine, visualization) => {
    const mutation = {
        type: 'create',
        resource: 'visualizations',
        data: visualization,
        params: {
            skipTranslations: 'true',
            skipSharing: 'true',
        },
    }

    if (visualization.id) {
        mutation.type = 'update'
        mutation.id = visualization.id
    }

    return engine.mutate(mutation)
}
