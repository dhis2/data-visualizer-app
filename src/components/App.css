/* App */

.data-visualizer-app {
    block-size: 100%;
    background-color: var(--colors-grey300);
    overflow: hidden;
}
.data-visualizer-app * {
    outline: none;
}

body {
    font-family: 'Roboto', sans-serif;
}

/* Flex */

.flex-ct {
    display: flex;
}

.flex-dir-col {
    flex-direction: column;
}

.flex-grow-1 {
    flex-grow: 1;
}

.flex-basis-0 {
    flex-basis: 0%;
}

/* Headerbar */

.section-headerbar {
    block-size: 48px;
}

/* Main */
.section-main {
    overflow: hidden;
}

.main-left {
    min-inline-size: 260px;
}

/* Main center */

.main-center-canvas {
    display: flex;
    justify-content: center;
    block-size: 100%;
    overflow: hidden;
    position: relative;
}

.main-center-layout {
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.03);
}
