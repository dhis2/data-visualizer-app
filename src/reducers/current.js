export const actionTypes = {
    SET_CURRENT: 'SET_CURRENT',
};

export const DEFAULT_CURRENT = {};

export default (state = DEFAULT_CURRENT, action) => {
    switch (action.type) {
        case actionTypes.SET_CURRENT: {
            return action.value;
        }
        default:
            return state;
    }
};

export const sGetFromState = state => state.current;
