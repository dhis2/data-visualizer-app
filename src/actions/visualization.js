import { actionTypes } from '../reducers';
import { apiFetchVisualization } from '../api/visualization';
import { acSetCurrent } from '../actions/current';

export const acSetVisualization = visualization => ({
    type: actionTypes.SET_VISUALIZATION,
    value: visualization,
});

export const tSetVisualization = (type, id) => async (dispatch, getState) => {
    const onSuccess = visualization => {
        dispatch(acSetVisualization(visualization));
        dispatch(acSetCurrent(visualization));
    };

    const onError = error => {
        console.log('Error (apiFetchVisualization): ', error);
        return error;
    };

    try {
        const response = await apiFetchVisualization(type, id);
        return onSuccess(response);
    } catch (err) {
        return onError(err);
    }
};
