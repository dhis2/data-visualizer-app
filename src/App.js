import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Snackbar from 'material-ui/Snackbar';
import i18n from '@dhis2/d2-i18n';

import SnackbarMessage from './widgets/SnackbarMessage';
import MenuBar from './MenuBar/MenuBar';
import VisualizationTypeSelector from './VisualizationTypeSelector/VisualizationTypeSelector';
import Dimensions from './Dimensions/Dimensions';
import Visualization from './Visualization/Visualization';
import * as fromReducers from './reducers';
import * as fromActions from './actions';

import './App.css';

export class App extends Component {
    componentDidMount() {
        const { store } = this.context;
        const d2 = this.props.d2;
        store.dispatch(fromActions.fromUser.acReceivedUser(d2.currentUser));
        store.dispatch(fromActions.fromDimensions.tSetDimensions());
    }

    getChildContext() {
        return {
            baseUrl: this.props.baseUrl,
            i18n,
            d2: this.props.d2,
        };
    }

    renderSnackbar() {
        return (
            <Snackbar
                open={this.props.snackbarOpen}
                message={
                    <SnackbarMessage message={this.props.snackbarMessage} />
                }
                autoHideDuration={this.props.snackbarDuration}
                onRequestClose={this.props.onCloseSnackbar}
            />
        );
    }

    render() {
        return (
            <Fragment>
                <div className="app">
                    <div className="item1 headerbar">Headerbar</div>
                    <div className="item2 visualization-type-selector">
                        <VisualizationTypeSelector />
                    </div>
                    <div className="item3 menu-bar">
                        <MenuBar />
                    </div>
                    <div className="item4 dimensions">
                        <Dimensions />
                    </div>
                    <div className="item5 chart-layout">Chart layout</div>
                    <div className="item6 interpretations">
                        Interpretations panel
                    </div>
                    <div className="item7 canvas">
                        <Visualization />
                    </div>
                </div>
                {this.renderSnackbar()}
            </Fragment>
        );
    }
}

const mapStateToProps = state => {
    const { message, duration, open } = fromReducers.fromSnackbar.sGetSnackbar(
        state
    );
    return {
        snackbarOpen: open,
        snackbarMessage: message,
        snackbarDuration: duration,
    };
};

App.contextTypes = {
    store: PropTypes.object,
};

App.childContextTypes = {
    d2: PropTypes.object,
    baseUrl: PropTypes.string,
    i18n: PropTypes.object,
};

App.propTypes = {
    d2: PropTypes.object,
    baseUrl: PropTypes.string,
};

export default connect(
    mapStateToProps,
    {
        onCloseSnackbar: fromActions.fromSnackbar.acCloseSnackbar,
    }
)(App);
