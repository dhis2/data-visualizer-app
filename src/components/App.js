import { useCachedDataQuery, Toolbar } from '@dhis2/analytics'
import { useSetting } from '@dhis2/app-service-datastore'
import i18n from '@dhis2/d2-i18n'
import {
    CssVariables,
    Modal,
    ModalTitle,
    ModalContent,
    ModalActions,
    ButtonStrip,
    Button,
} from '@dhis2/ui'
import React, { useEffect, useCallback, useRef, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {
    acClearCurrent,
    acSetCurrentFromUi,
    tSetCurrentFromUi,
} from '../actions/current.js'
import { tSetDimensions } from '../actions/dimensions.js'
import { clearAll, tDoLoadVisualization } from '../actions/index.js'
import { acAddMetadata } from '../actions/metadata.js'
import { acAddSettings } from '../actions/settings.js'
import { acAddParentGraphMap, acSetUiFromVisualization } from '../actions/ui.js'
import { acReceivedUser, tLoadUserAuthority } from '../actions/user.js'
import { acClearVisualization } from '../actions/visualization.js'
import { Snackbar } from '../components/Snackbar/Snackbar.js'
import { USER_DATASTORE_CURRENT_AO_KEY } from '../modules/currentAnalyticalObject.js'
import history from '../modules/history.js'
import defaultMetadata from '../modules/metadata.js'
import { getParentGraphMapFromVisualization } from '../modules/ui.js'
import { STATE_DIRTY, getVisualizationState } from '../modules/visualization.js'
import { sGetCurrent } from '../reducers/current.js'
import { sGetUi } from '../reducers/ui.js'
import { sGetVisualization } from '../reducers/visualization.js'
import { default as DetailsPanel } from './DetailsPanel/DetailsPanel.js'
import DimensionsPanel from './DimensionsPanel/DimensionsPanel.js'
import DndContext from './DndContext.js'
import { InterpretationModal } from './InterpretationModal/index.js'
import Layout from './Layout/Layout.js'
import { MenuBar } from './MenuBar/MenuBar.js'
import { TitleBar } from './TitleBar/TitleBar.js'
import { Visualization } from './Visualization/Visualization.js'
import { APPROVAL_LEVEL_OPTION_AUTH } from './VisualizationOptions/Options/ApprovalLevel.js'
import { VisualizationTypeSelector } from './VisualizationTypeSelector/VisualizationTypeSelector.js'
import './App.css'
import './scrollbar.css'

const App = () => {
    const [currentAO] = useSetting(USER_DATASTORE_CURRENT_AO_KEY)

    const [previousLocation, setPreviousLocation] = useState(null)
    const [initialLoadIsComplete, setInitialLoadIsComplete] = useState(false)
    const [locationToConfirm, setLocationToConfirm] = useState(false)

    const dispatch = useDispatch()

    const current = useSelector(sGetCurrent)
    const ui = useSelector(sGetUi)
    const visualization = useSelector(sGetVisualization)

    const { currentUser, systemSettings, orgUnitLevels, rootOrgUnits } =
        useCachedDataQuery()

    const interpretationsUnitRef = useRef()
    const onInterpretationUpdate = () => {
        interpretationsUnitRef.current.refresh()
    }

    const parseLocation = (location) => {
        const pathParts = location.pathname.slice(1).split('/')
        const id = pathParts[0]
        const interpretationId = pathParts[2]
        return { id, interpretationId }
    }

    const loadVisualization = useCallback(
        (location) => {
            /**
             * The following cases require a fetch/refetch of the AO
             * - enter a new url (causing a page load)
             * - file->open (same or different AO)
             * - file->saveAs
             */
            const isRefetchNeeded = (location) => {
                if (!previousLocation) {
                    return true
                }

                const id = location.pathname.slice(1).split('/')[0]
                const prevId = previousLocation.slice(1).split('/')[0]

                if (id !== prevId || previousLocation === location.pathname) {
                    return true
                }

                return false
            }

            if (location.pathname.length > 1) {
                // /currentAnalyticalObject
                // /${id}/
                // /${id}/interpretation/${interpretationId}
                const { id } = parseLocation(location)

                const urlContainsCurrentAOKey =
                    id === USER_DATASTORE_CURRENT_AO_KEY

                if (urlContainsCurrentAOKey) {
                    dispatch(
                        acAddParentGraphMap(
                            getParentGraphMapFromVisualization(currentAO)
                        )
                    )

                    // clear visualization and current
                    // to avoid leave them "dirty" when navigating to
                    // /currentAnalyticalObject from a previously saved AO
                    dispatch(acClearVisualization())
                    dispatch(acClearCurrent())

                    dispatch(acSetUiFromVisualization(currentAO))
                    dispatch(tSetCurrentFromUi())
                }

                if (!urlContainsCurrentAOKey && isRefetchNeeded(location)) {
                    dispatch(
                        tDoLoadVisualization({
                            id,
                            ouLevels: orgUnitLevels,
                        })
                    )
                }
            } else {
                dispatch(clearAll()) // XXX
            }
            setInitialLoadIsComplete(true)
            setPreviousLocation(location.pathname)
        },
        [
            currentAO,
            dispatch,
            orgUnitLevels,
            previousLocation,
            setPreviousLocation,
        ]
    )

    useEffect(
        () => {
            dispatch(
                // XXX see how to write this better
                acAddSettings({
                    ...systemSettings,
                    uiLocale: currentUser.settings.uiLocale,
                    displayProperty: currentUser.settings.displayProperty,
                    displayNameProperty:
                        currentUser.settings.displayNameProperty,
                    rootOrganisationUnits: rootOrgUnits,
                })
            )
            dispatch(tLoadUserAuthority('ALL'))
            dispatch(tLoadUserAuthority(APPROVAL_LEVEL_OPTION_AUTH))
            dispatch(acReceivedUser(currentUser))
            dispatch(tSetDimensions())

            const metaData = { ...defaultMetadata() }

            rootOrgUnits.forEach((rootOrgUnit) => {
                if (rootOrgUnit.id) {
                    metaData[rootOrgUnit.id] = {
                        ...rootOrgUnit,
                        path: `/${rootOrgUnit.id}`,
                    }
                }
            })

            dispatch(acAddMetadata(metaData))

            loadVisualization(history.location)

            const unlisten = history.listen(({ location }) => {
                const isSaving = location.state?.isSaving
                const isOpening = location.state?.isOpening
                const isResetting = location.state?.isResetting
                const isModalOpening = location.state?.isModalOpening
                const isModalClosing = location.state?.isModalClosing
                const isValidLocationChange =
                    previousLocation !== location.pathname &&
                    !isModalOpening &&
                    !isModalClosing

                if (
                    // currently editing
                    getVisualizationState(visualization, current) ===
                        STATE_DIRTY &&
                    // wanting to navigate elsewhere
                    previousLocation !== location.pathname &&
                    // not saving
                    !isSaving
                ) {
                    setLocationToConfirm(location)
                } else {
                    if (
                        isSaving ||
                        isOpening ||
                        isResetting ||
                        isValidLocationChange
                    ) {
                        loadVisualization(location)
                    }

                    setLocationToConfirm(null)
                }
            })

            document.body.addEventListener(
                'keyup',
                (e) =>
                    e.key === 'Enter' &&
                    e.ctrlKey === true &&
                    dispatch(acSetCurrentFromUi(ui))
            )

            window.addEventListener('beforeunload', (event) => {
                if (
                    getVisualizationState(visualization, current) ===
                    STATE_DIRTY
                ) {
                    event.preventDefault()
                    event.returnValue = i18n.t('You have unsaved changes.')
                }
            })

            return () => unlisten && unlisten()
        },
        [
            //        current,
            //        currentUser,
            //        dispatch,
            //        loadVisualization,
            //        previousLocation,
            //        rootOrgUnits,
            //        systemSettings,
            //        ui,
            //        visualization,
        ]
    )

    // TODO continue from here

    //        this.unlisten = history.listen(({ location }) => {
    //            const isSaving = location.state?.isSaving
    //            const isOpening = location.state?.isOpening
    //            const isResetting = location.state?.isResetting
    //            /*
    //            const isModalOpening = location.state?.isModalOpening
    //            const isModalClosing = location.state?.isModalClosing
    //            const isValidLocationChange =
    //                this.state.previousLocation !== location.pathname &&
    //                !isModalOpening &&
    //                !isModalClosing
    //*/
    //            if (
    //                // currently editing
    //                getVisualizationState(
    //                    this.props.visualization,
    //                    this.props.current
    //                ) === STATE_DIRTY &&
    //                // wanting to navigate elsewhere
    //                this.state.previousLocation !== location.pathname &&
    //                // not saving
    //                !isSaving
    //            ) {
    //                this.setState({ locationToConfirm: location })
    //            } else {
    //                if (
    //                    isSaving ||
    //                    isOpening ||
    //                    isResetting ||
    //                    this.state.previousLocation !== location.pathname
    //                ) {
    //                    this.loadVisualization(location)
    //                }
    //
    //                this.setState({ locationToConfirm: null })
    //            }
    //        })

    return (
        <>
            <div className="data-visualizer-app flex-ct flex-dir-col">
                <Toolbar>
                    <VisualizationTypeSelector />
                    <MenuBar
                        apiObjectName="visualization"
                        dataTest={'app-menubar'}
                    />
                </Toolbar>
                <div className="section-main flex-grow-1 flex-ct">
                    <DndContext>
                        <div className="main-left">
                            <DimensionsPanel />
                        </div>
                        <div className="main-center flex-grow-1 flex-basis-0 flex-ct flex-dir-col">
                            <div className="main-center-layout">
                                <Layout />
                            </div>
                            <div className="main-center-titlebar">
                                <TitleBar />
                            </div>
                            <div className="main-center-canvas flex-grow-1">
                                {initialLoadIsComplete && <Visualization />}
                                {current && (
                                    <InterpretationModal
                                        onInterpretationUpdate={
                                            onInterpretationUpdate
                                        }
                                    />
                                )}
                            </div>
                        </div>
                    </DndContext>
                    {ui.rightSidebarOpen && current && (
                        <div className="main-right">
                            <DetailsPanel
                                interpretationsUnitRef={interpretationsUnitRef}
                            />
                        </div>
                    )}
                </div>
            </div>
            {locationToConfirm && (
                <Modal small dataTest={'confirm-leave-modal'}>
                    <ModalTitle>
                        {i18n.t('Discard unsaved changes?')}
                    </ModalTitle>
                    <ModalContent>
                        {i18n.t(
                            'Are you sure you want to leave this visualization? Any unsaved changes will be lost.'
                        )}
                    </ModalContent>
                    <ModalActions>
                        <ButtonStrip end>
                            <Button
                                secondary
                                onClick={() => {
                                    setLocationToConfirm(null), history.back()
                                }}
                                dataTest={'confirm-leave-modal-option-cancel'}
                            >
                                {i18n.t('No, cancel')}
                            </Button>

                            <Button
                                onClick={() => {
                                    loadVisualization(locationToConfirm)

                                    setLocationToConfirm(null)
                                }}
                                primary
                                dataTest={'confirm-leave-modal-option-confirm'}
                            >
                                {i18n.t('Yes, leave')}
                            </Button>
                        </ButtonStrip>
                    </ModalActions>
                </Modal>
            )}
            <Snackbar />
            <CssVariables colors spacers elevations />
        </>
    )
}

export default App
