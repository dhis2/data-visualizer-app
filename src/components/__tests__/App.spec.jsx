import React from 'react'
import { waitFor, act } from '@testing-library/react'
import { renderWithProviders } from '../../../config/testsContext.js'
// import { USER_DATASTORE_CURRENT_AO_KEY } from '../../modules/currentAnalyticalObject.js'
// import history from '../../modules/history.js'
// import * as ui from '../../modules/ui.js'
import { App } from '../App.jsx'
import * as fromActions from '../../actions/index.js'
import history from '../../modules/history.js'
import { DEFAULT_CURRENT } from '../../reducers/current.js'
import { setupTestStore } from '../../configureStore.js'
import { apiFetchVisualization } from '../../api/visualization.js'
import { DEFAULT_VISUALIZATION } from '../../reducers/visualization.js'
import { useSetting } from '@dhis2/app-service-datastore'
import { USER_DATASTORE_CURRENT_AO_KEY } from '../../modules/currentAnalyticalObject.js'

jest.mock('@dhis2/analytics', () => ({
    ...jest.requireActual('@dhis2/analytics'),
    // apiFetchOrganisationUnitLevels: jest.fn(),
    // getPredefinedDimensions: () => {},
    visTypeDisplayNames: {},
    Toolbar: () => <div>Toolbar</div>,
    useCachedDataQuery: jest.fn(() => ({
        currentUser: { id: 'user1', settings: {} },
        orgUnitLevels: [],
        rootOrgUnits: [],
        systemSettings: {},
    })),
    convertOuLevelsToUids: jest.fn((ouLevels, visualization) => visualization),
    apiFetchDimensions: jest.fn(() =>
        Promise.resolve([
            { id: 'dim1', name: 'Dimension 1' },
            { id: 'dim2', name: 'Dimension 2' },
            { id: 'dim3', name: 'Dimension 3' },
        ])
    ),
}))

jest.mock('@dhis2/app-service-datastore', () => ({
    ...jest.requireActual('@dhis2/app-service-datastore'),
    useSetting: jest.fn(() => [{}]), // default: empty object
}))

jest.mock('../../api/visualization.js', () => ({
    ...jest.requireActual('../../api/visualization.js'),
    apiFetchVisualization: jest.fn(() =>
        Promise.resolve({
            visualization: {},
        })
    ),
}))

jest.mock('../../api/dataStatistics.js', () => ({
    ...jest.requireActual('../../api/dataStatistics.js'),
    apiPostDataStatistics: jest.fn(() => Promise.resolve({ result: 'ok' })),
}))

jest.mock('../DndContext.jsx', () => () => <div>DndContext</div>)
jest.mock('../DetailsPanel/DetailsPanel.jsx', () => () => (
    <div>DetailsPanel</div>
))

jest.mock('@dhis2/ui', () => {
    const original = jest.requireActual('@dhis2/ui')
    return {
        ...original,
        Modal: (props) => <div {...props}>Modal</div>,
    }
})

jest.mock('@dhis2/app-runtime', () => ({
    useDataEngine: jest.fn(() => ({
        query: Function.prototype,
    })),
}))

jest.mock('../Visualization/Visualization.jsx', () => ({
    __esModule: true,
    Visualization: () => <div />,
}))

describe('App', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })
    test('renders the app skeleton', async () => {
        const reduxState = {
            visualization: { name: 'Hello', id: 'hello-id' },
            current: { name: 'Hello', id: 'hello-id' },
            ui: { rightSidebarOpen: false },
        }

        Object.defineProperty(history, 'location', {
            configurable: true,
            get: () => ({
                pathname: '/',
                search: '',
                state: {},
                hash: '',
                key: 'test-key',
            }),
        })

        const store = setupTestStore(reduxState)
        jest.spyOn(store, 'dispatch')
        const { container } = renderWithProviders(<App />, store)

        await waitFor(() => {
            expect(container).toHaveTextContent('Toolbar')
        })

        expect(container).toMatchSnapshot()

        // TODO
        // const calls = store.dispatch.mock.calls.map(([action]) => action)
        // console.log('Dispatched actions:', calls)

        // const wasClearVisualizationDispatched = calls.some(
        //     (action) =>
        //         typeof action === 'object' &&
        //         action.type ===
        //             fromActions.fromVisualization.acClearVisualization().type
        // )
        // expect(wasClearVisualizationDispatched).toBe(true)

        // const wasClearCurrentDispatched = calls.some(
        //     ([action]) =>
        //         typeof action === 'object' &&
        //         action.type === fromActions.fromCurrent.acClearCurrent().type
        // )
        // expect(wasClearCurrentDispatched).toBe(true)
    })

    test('calls setVisualization when location pathname has length', async () => {
        const reduxState = {
            loader: {
                isLoading: false,
                loadingError: 'there was an error',
                isPluginLoading: true,
            },
        }

        const newVisualization = {
            id: 'goodbye-id',
            name: 'Goodbye',
            columns: [{ dimension: 'dx', items: [{ id: 'anc1' }] }],
            rows: [{ dimension: 'ou', items: [{ id: 'myou1' }] }],
            filters: [{ dimension: 'pe', items: [{ id: '202301' }] }],
        }

        apiFetchVisualization.mockResolvedValueOnce({
            visualization: newVisualization,
        })

        Object.defineProperty(history, 'location', {
            get: () => ({
                pathname: '/goodbyeuid',
            }),
        })

        const store = setupTestStore(reduxState)

        expect(store.getState().visualization).toEqual(DEFAULT_VISUALIZATION)
        expect(store.getState().current).toEqual(DEFAULT_CURRENT)
        expect(store.getState().loader).toEqual({
            isLoading: false,
            loadingError: 'there was an error',
            isPluginLoading: true,
        })
        const { container } = renderWithProviders(<App />, store)

        await waitFor(() => {
            expect(container).toHaveTextContent('Toolbar')
        })

        expect(apiFetchVisualization).toHaveBeenCalled()

        expect(store.getState().visualization).toEqual(newVisualization)
        expect(store.getState().current).toEqual(newVisualization)
        expect(store.getState().loader).toEqual({
            isLoading: false,
            loadingError: null,
            isPluginLoading: true,
        })
    })

    test('resets state when location pathname is root', async () => {
        const reduxState = {
            visualization: { name: 'Hello', id: 'hello-id' },
            current: { name: 'Hello', id: 'hello-id' },
            ui: { rightSidebarOpen: true },
        }

        Object.defineProperty(history, 'location', {
            configurable: true,
            get: () => ({
                pathname: '/',
                search: '',
                state: {},
                hash: '',
                key: 'test-key',
            }),
        })

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        await waitFor(() => {
            expect(store.getState().visualization).toEqual(null)
            expect(store.getState().current).toEqual(null)
            expect(store.getState().ui.rightSidebarOpen).toBe(false)
        })
    })

    test('loads AO from user data store if id equals to "currentAnalyticalObject"', async () => {
        // Arrange: set up the AO object to be returned by useSetting
        const aoFromUserDataStore = {
            columns: [{ dimension: 'dx', items: [{ id: 'anc1' }] }],
            rows: [{ dimension: 'ou', items: [{ id: 'myou1' }] }],
            filters: [{ dimension: 'pe', items: [{ id: '202301' }] }],
        }
        useSetting
            .mockImplementationOnce(() => [aoFromUserDataStore]) // First call
            .mockImplementationOnce(() => [aoFromUserDataStore]) // Second call

        // Set up initial redux state
        const reduxState = {
            visualization: { name: 'Hello', id: 'hello-id' },
            current: { name: 'Hello', id: 'hello-id' },
            ui: { rightSidebarOpen: true },
        }

        // Mock the location to simulate the special AO key
        Object.defineProperty(history, 'location', {
            get: () => ({
                pathname: `/${USER_DATASTORE_CURRENT_AO_KEY}`,
            }),
        })

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        expect(store.getState().visualization).toEqual(null)
        expect(store.getState().current).toEqual(
            expect.objectContaining(aoFromUserDataStore)
        )
        expect(store.getState().ui.rightSidebarOpen).toBe(true)
        expect(store.getState().ui.itemsByDimension).toEqual({
            dx: ['anc1'],
            ou: ['myou1'],
            pe: ['202301'],
        })
    })

    test('reloads visualization when opening the same visualization', async () => {
        // Initial redux state
        const reduxState = {
            visualization: { id: 'fluttershy', name: 'Fluttershy' },
            current: { id: 'fluttershy', name: 'Fluttershy' },
            ui: { rightSidebarOpen: false },
        }

        // Set initial location
        Object.defineProperty(history, 'location', {
            configurable: true,
            get: () => ({
                pathname: '/fluttershy',
                search: '',
                state: {},
                hash: '',
                key: 'test-key',
            }),
        })

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        // Wait for initial visualization load
        await waitFor(() => {
            expect(store.getState().visualization).toEqual(
                expect.objectContaining({ id: 'fluttershy' })
            )
        })

        // Simulate opening the same visualization again
        await act(async () => {
            history.replace({ pathname: '/fluttershy' }, { isOpening: true })
        })

        // Wait for the visualization to reload (should be called again)
        await waitFor(() => {
            expect(apiFetchVisualization).toHaveBeenCalledTimes(2)
        })
    })

    test('does not reload visualization when interpretation toggled', async () => {
        // Initial redux state
        const reduxState = {
            visualization: { id: 'applejack', name: 'Applejack' },
            current: { id: 'applejack', name: 'Applejack' },
            ui: { rightSidebarOpen: false },
        }

        // Set initial location to interpretation route
        Object.defineProperty(history, 'location', {
            get: () => ({
                pathname: '/applejack/interpretation/xyz123',
            }),
        })

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        // Wait for initial visualization load
        await waitFor(() => {
            expect(store.getState().visualization).toEqual(
                expect.objectContaining({ id: 'applejack' })
            )
        })

        // Simulate toggling interpretation (navigating to base visualization route)
        await act(async () => {
            history.push('/applejack')
        })

        // The visualization should NOT reload (apiFetchVisualization should only be called once)
        expect(apiFetchVisualization).toHaveBeenCalledTimes(1)
    })

    test('reloads visualization when same pathname pushed when saving', async () => {
        // Initial redux state
        const reduxState = {
            visualization: { id: 'fluttershy', name: 'Fluttershy' },
            current: { id: 'fluttershy', name: 'Fluttershy' },
            ui: { rightSidebarOpen: false },
        }

        // Set initial location
        Object.defineProperty(history, 'location', {
            get: () => ({
                pathname: '/fluttershy',
            }),
        })

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        // Wait for initial visualization load
        await waitFor(() => {
            expect(store.getState().visualization).toEqual(
                expect.objectContaining({ id: 'fluttershy' })
            )
        })

        // Simulate pushing the same pathname with isSaving flag
        await act(async () => {
            history.replace({ pathname: '/fluttershy' }, { isSaving: true })
        })

        // The visualization should reload (apiFetchVisualization should be called twice)
        expect(apiFetchVisualization).toHaveBeenCalledTimes(2)
    })

    test('triggers popstate events on history push events', async () => {
        // Set initial location
        Object.defineProperty(history, 'location', {
            configurable: true,
            get: () => ({
                pathname: '/rarity',
                search: '',
                state: {},
                hash: '',
                key: 'test-key',
            }),
        })

        const reduxState = {
            visualization: { id: 'rarity', name: 'Rarity' },
            current: { id: 'rarity', name: 'Rarity' },
            ui: { rightSidebarOpen: false },
        }

        const popstateHandler = jest.fn()
        window.addEventListener('popstate', popstateHandler)

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        // Simulate history.push
        await act(async () => {
            history.push('/rainbowdash')
        })

        expect(popstateHandler).toBeCalledTimes(1)

        window.removeEventListener('popstate', popstateHandler)
    })

    test('triggers popstate events on history replace events', async () => {
        // Set initial location
        Object.defineProperty(history, 'location', {
            configurable: true,
            get: () => ({
                pathname: '/rarity',
                search: '',
                state: {},
                hash: '',
                key: 'test-key',
            }),
        })

        const reduxState = {
            visualization: { id: 'rarity', name: 'Rarity' },
            current: { id: 'rarity', name: 'Rarity' },
            ui: { rightSidebarOpen: false },
        }

        const popstateHandler = jest.fn()
        window.addEventListener('popstate', popstateHandler)

        const store = setupTestStore(reduxState)
        renderWithProviders(<App />, store)

        // Simulate history.replace
        await act(async () => {
            history.replace('/rainbowdash')
        })

        expect(popstateHandler).toBeCalledTimes(1)

        window.removeEventListener('popstate', popstateHandler)
    })
})
