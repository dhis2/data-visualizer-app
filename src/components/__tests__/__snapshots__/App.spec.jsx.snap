// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the app skeleton 1`] = `
<div>
  <div
    class="data-visualizer-app flex-ct flex-dir-col"
  >
    <div>
      Toolbar
    </div>
    <div
      class="section-main flex-grow-1 flex-ct"
    >
      <div>
        DndContext
      </div>
    </div>
  </div>
</div>
`;